httpx/auth.py

```diff
@@ -36,3 +40,132 @@ def build_auth_header(self) -> str:
         userpass = b":".join((username, password))
         token = b64encode(userpass).decode().strip()
         return f"Basic {token}"
+
+
+class HTTPDigestAuth:
+
+    ALGORITHM_TO_HASH_FUNCTION = {
+        "MD5": hashlib.md5,
+        "SHA": hashlib.sha1,
+        "SHA-256": hashlib.sha256,
+        "SHA-512": hashlib.sha512,
+    }
+
+    def __init__(
+        self, username: typing.Union[str, bytes], password: typing.Union[str, bytes]
+    ) -> None:
+        self.username = username
+        self.password = password
+
+    def __call__(self, request: AsyncRequest, response) -> AsyncRequest:
+        auth_header = self.build_auth_header(request, response)
+        if auth_header is not None:
+            # TODO: raise instead of returning None?
+            request.headers["Authorization"] = auth_header
+        return request
+
+    def build_auth_header(
+        self, request: AsyncRequest, response: AsyncResponse
+    ) -> typing.Optional[str]:
+        # Retrieve challenge from response header
+        header = response.headers.get("www-authenticate")
+        assert header.lower().startswith("digest")
+        challenge = self._parse_header(header)
+        algorithm = challenge.get("algorithm", "MD5")
+
+        realm = _safe_encode(challenge["realm"])
+        nonce = _safe_encode(challenge["nonce"])
+        qop = _safe_encode(challenge["qop"]) if "qop" in challenge else None
+        opaque = _safe_encode(challenge["opaque"]) if "opaque" in challenge else None
+        username = _safe_encode(self.username)
+        password = _safe_encode(self.password)
+
+        # Assemble parts depending on hash algorithms
+        hash_func = self.ALGORITHM_TO_HASH_FUNCTION[
+            algorithm.replace("-SESS", "").upper()
+        ]
+
+        def digest(data: bytes) -> bytes:
+            return hash_func(data).hexdigest().encode("utf-8")
+
+        def keyed_digest(secret, data) -> bytes:
+            return digest(b":".join((secret, data)))
+
+        A1 = b":".join((username, realm, password))
+        HA1 = digest(A1)
+
+        path = self._get_parsed_path(request.url)
+        A2 = b":".join((request.method.encode("utf-8"), path))
+        # TODO: implement auth-int
+        HA2 = digest(A2)
+
+        # Construct Authenticate header string
+        nonce_count = 1  # TODO: use contextvars to count properly
```

Comment:
I'm not familiar at all with `contextvars` but from what I could gather from the docs the current context if copied when creating an asyncio Task which *should* match our requirements for this?

Reply:
No - wouldn't want anything here that's coupled up to concurrency primitives.
Either an incrementing instance attribute, or an incrementing class attribute would be preferable. (I'm not sure which of those two is appropriate without digging into exactly how the nonce count is used)

Reply:
Right, makes sense. [Requests uses thread locals extensively to hold state](https://github.com/not-kennethreitz/requests/blob/9a2e5df000691ad28613524ca7f80aa28a94a8d5/requests/auth.py#L117-L125) which I naively thought we should carry over.